File: backend/app/schemas/config_sections/game_config.py
from typing import Optional
from pydantic import BaseModel, Field


class Vector2(BaseModel):
    """2D Vector for coordinate pairs"""
    x: float = Field(..., description="X coordinate")
    y: float = Field(..., description="Y coordinate")
    


class GameLogicConfig(BaseModel):
    """Game logic configuration subsection"""
    matchCount: int = Field(..., ge=1, description="Number of tiles needed to match")
    countUndoTileRevive: int = Field(..., ge=0, description="Count of undo tile revives")
    countShuffleTileRevive: int = Field(..., ge=0, description="Count of shuffle tile revives")
    countSlotHolder: int = Field(..., ge=1, description="Number of slot holders")
    warningThreshold: int = Field(..., ge=0, description="Warning threshold value")
    


class ComboConfig(BaseModel):
    """Combo configuration subsection"""
    matchEffect: int = Field(..., ge=0, description="Match effect value")
    maxNoMatch: int = Field(..., ge=0, description="Maximum no match count")
    


class GameLogic(BaseModel):
    """Game logic section containing game logic config and combo"""
    gameLogicConfig: GameLogicConfig = Field(..., description="Game logic configuration")
    combo: ComboConfig = Field(..., description="Combo configuration")
    


class GridView(BaseModel):
    """Grid view configuration"""
    tileSize: Vector2 = Field(..., description="Size of tiles")
    


class HolderView(BaseModel):
    """Holder view configuration"""
    slotSize: Vector2 = Field(..., description="Size of slots")
    slotSpace: float = Field(..., ge=0, description="Space between slots")
    ratioBetweenTwoTile: float = Field(..., ge=0, description="Ratio between two tiles")
    slotYPadding: float = Field(..., ge=0, description="Slot Y padding")
    tileInHolderYPadding: float = Field(..., ge=0, description="Tile in holder Y padding")
    


class ViewConfig(BaseModel):
    """View configuration section containing grid and holder views"""
    gridView: GridView = Field(..., description="Grid view configuration")
    holderView: HolderView = Field(..., description="Holder view configuration")
    


class GameConfig(BaseModel):
    """Game configuration - logic and view settings"""
    gameLogic: GameLogic = Field(..., description="Game logic settings")
    viewConfig: ViewConfig = Field(..., description="View configuration settings")
    

File: frontend/src/lib/validations/gameConfig.ts
import { z } from 'zod';

// Vector2 type for X/Y coordinate pairs
const vector2Schema = z.object({
  x: z.number({ required_error: 'X value is required' }),
  y: z.number({ required_error: 'Y value is required' }),
});

// Game Logic Config subsection
const gameLogicConfigSchema = z.object({
  matchCount: z.number({ required_error: 'Match Count is required' }).int().min(1),
  countUndoTileRevive: z.number({ required_error: 'Count Undo Tile Revive is required' }).int().min(0),
  countShuffleTileRevive: z.number({ required_error: 'Count Shuffle Tile Revive is required' }).int().min(0),
  countSlotHolder: z.number({ required_error: 'Count Slot Holder is required' }).int().min(1),
  warningThreshold: z.number({ required_error: 'Warning Threshold is required' }).int().min(0),
});

// Combo subsection
const comboSchema = z.object({
  matchEffect: z.number({ required_error: 'Match Effect is required' }).int().min(0),
  maxNoMatch: z.number({ required_error: 'Max No Match is required' }).int().min(0),
});

// Grid View subsection
const gridViewSchema = z.object({
  tileSize: vector2Schema,
});

// Holder View subsection
const holderViewSchema = z.object({
  slotSize: vector2Schema,
  slotSpace: z.number({ required_error: 'Slot Space is required' }).min(0),
  ratioBetweenTwoTile: z.number({ required_error: 'Ratio Between Two Tile is required' }).min(0),
  slotYPadding: z.number({ required_error: 'Slot Y Padding is required' }).min(0),
  tileInHolderYPadding: z.number({ required_error: 'Tile In Holder Y Padding is required' }).min(0),
});

// Game Logic section (contains Game Logic Config + Combo)
const gameLogicSchema = z.object({
  gameLogicConfig: gameLogicConfigSchema,
  combo: comboSchema,
});

// View Config section (contains Grid View + Holder View)
const viewConfigSchema = z.object({
  gridView: gridViewSchema,
  holderView: holderViewSchema,
});

// Full Game Config
export const gameConfigSchema = z.object({
  gameLogic: gameLogicSchema,
  viewConfig: viewConfigSchema,
});

// Type exports
export type Vector2 = z.infer<typeof vector2Schema>;
export type GameLogicConfig = z.infer<typeof gameLogicConfigSchema>;
export type Combo = z.infer<typeof comboSchema>;
export type GridView = z.infer<typeof gridViewSchema>;
export type HolderView = z.infer<typeof holderViewSchema>;
export type GameLogic = z.infer<typeof gameLogicSchema>;
export type ViewConfig = z.infer<typeof viewConfigSchema>;
export type GameConfig = z.infer<typeof gameConfigSchema>;

